packages/core/src/login.ts

```diff
@@ -0,0 +1,118 @@
+          'device_code': string
+          'user_code': string
+          'verification_uri': string
+          'expires_in': number
+          'interval': number
+          'verification_uri_complete': string
+        }
+
+  exec(`open ${responseData.verification_uri_complete}`) // TODO - use cross-platform, safer, opener - once we get esm modules working
+  logger.info(`If your browser did not open, here: ${responseData.verification_uri_complete}`)
+  logger.info('Make sure code is ', responseData.user_code)
+
+  return withSpinner(() => new Promise<TokesFileSchema>((resolve, reject) => {
+    const intervalId = setInterval(async () => {
+      try {
+        const tokenResponse = await fetch(`${loginUrl}/oauth/token`, { method: 'POST',
+          headers: { 'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8' },
+          body: new URLSearchParams({ grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
+            device_code: responseData.device_code,
+            client_id: 'jEnySAwuAaWaLOdWdALbvXj6dZEqgAJB' }) })
+
+        if (!tokenResponse.ok && tokenResponse.status >= 500) throw new Error(`Bad response from token endpoint: ${tokenResponse.status}: ${tokenResponse.statusText}`)
+
+        const tokenResponseData = await tokenResponse.json()
+        if ('access_token' in tokenResponseData) {
+          clearInterval(intervalId)
+          const tokensFile: TokesFileSchema = {
+            access_token: tokenResponseData.access_token,
+            id_token: tokenResponseData.id_token,
+          }
+          await fs.write(PERSISTENT_TOKEN_FILE_NAME, JSON.stringify(tokensFile))
+
+          resolve(tokenResponseData)
+        }
+      } catch (e) {
+        logger.info('Error getting tokens', e)
+        reject(e)
+      }
+    }, responseData.interval * 1000)
+  }), { opPrefix: 'Waiting for approval', successText: 'Done!' })
+}
+
+export const getTokens = async (dataDir: string) : Promise<TokesFileSchema | undefined> => {
+  const fs = localFs(dataDir)
+  const tokensFile = await fs.read(PERSISTENT_TOKEN_FILE_NAME)
+  if (tokensFile === undefined) return undefined
+
+  const tokens: TokesFileSchema = JSON.parse(tokensFile.toString())
+  const accessToken = jose.decodeJwt(tokens.access_token)
+  if (accessToken.exp === undefined || (accessToken.exp < Math.floor(Date.now() / 1000))) {
+    throw new TokenExpiredError()
+  }
+  return tokens
+}
+
+export const login = async (dataDir: string, loginUrl: string, lcUrl: string, logger: Logger) => {
+  let tokens: TokesFileSchema
+  try {
+    const tokensMaybe = await getTokens(dataDir)
+    if (tokensMaybe !== undefined) {
+      logger.info(`Already logged in as: ${jose.decodeJwt(tokensMaybe.id_token).email} 👌`)
+      return
+    }
+    tokens = await deviceFlow(dataDir, loginUrl, logger)
+  } catch (e) {
+    if (e instanceof TokenExpiredError) {
+      tokens = await deviceFlow(dataDir, loginUrl, logger)
+    } else {
+      throw e
```

Comment:
style nit, avoid else
```suggestion
    if (!(e instanceof TokenExpiredError)) {
      throw e 
    }
    tokens = await deviceFlow(dataDir, loginUrl, logger)
```